app/models/questionnaire.rb

```diff
@@ -0,0 +1,55 @@
+class Questionnaire < ApplicationRecord
+  has_many :questions, dependent: :restrict_with_error
+  belongs_to :instructor
+
+  before_destroy :check_for_question_associations
+
+  validate :validate_questionnaire
+  validates :name, presence: true
+  validates :max_question_score, :min_question_score, numericality: true
+
+  # clones the contents of a questionnaire, including the questions and associated advice
+  def self.copy_questionnaire_details(params)
+    orig_questionnaire = Questionnaire.find(params[:id])
+    questions = Question.where(questionnaire_id: params[:id])
+    questionnaire = orig_questionnaire.dup
+    questionnaire.name = 'Copy of ' + orig_questionnaire.name
+    questionnaire.created_at = Time.zone.now
+    questionnaire.updated_at = Time.zone.now
+    questionnaire.save!
+    questions.each do |question|
+      new_question = question.dup
+      new_question.questionnaire_id = questionnaire.id
+      new_question.save!
+    end
+    questionnaire
+  end
+
+  # validate the entries for this questionnaire
+  def validate_questionnaire
```

Comment:
The code looks good, but you can extract the validate_questionnaire method to a separate validator file to follow the Single Responsibility Principle. Here's how you can do it:

Create a new file called `questionnaire_validator.rb` in app/validators directory with the following code:

```ruby
class QuestionnaireValidator < ActiveModel::Validator
  def validate(record)
    record.errors.add(:max_question_score, 'The maximum question score must be a positive integer.') if record.max_question_score < 1
    record.errors.add(:min_question_score, 'The minimum question score must be a positive integer.') if record.min_question_score < 0
    record.errors.add(:min_question_score, 'The minimum question score must be less than the maximum.') if record.min_question_score >= record.max_question_score
    results = Questionnaire.where('id <> ? and name = ? and instructor_id = ?', record.id, record.name, record.instructor_id)
    record.errors.add(:name, 'Questionnaire names must be unique.') if results.present?
  end
end
```

Then, in your `questionnaire.rb` model file, update the validate_questionnaire method to use this validator:
```ruby
class Questionnaire < ApplicationRecord
  validates_with QuestionnaireValidator
end
```

Reply:
We tried this implementation for validations but the method call does not work implicitly. We would need extra time to understand the issue and refactor this.

Reply:
Are you attempting to use this validation in multiple instances and are wanting to call the specific method `validate`?
I just want to let you know that when you use `validates_with QuestionnaireValidator`, it automatically calls `validate(record)`, so you don't need to call this method explicitly.